src/tracer.ts

```diff
@@ -130,7 +151,10 @@ export default class Tracer {
      * @param  {any} carrier - see the documentation for the chosen `format`
      *         for a description of the carrier object.
      */
-    inject(spanContext, format, carrier) {
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_TEXT_MAP, carrier: { [key: string]: string }): void;
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_BINARY, carrier: { buffer: number[] }): void;
+    inject(spanContext: SpanContext | Span, format: typeof Constants.FORMAT_HTTP_HEADERS, carrier: { [key: string]: string }): void;
+    inject(spanContext: SpanContext | Span, format: Format, carrier: any): void {
```

Comment:
TypeScript does support both generic classes and methods. What I wanted to accomplish here was to enforce that if you eg pass `FORMAT_BINARY`, the `carrier` needs to be an object with a `buffer` property.
Overloads in TypeScript are not matched automatically, you just define the overloaded signatures, and then you must define one "generic" implementation that usually does runtime type checks. Implementations do not need to implement the 4 overloads.